### PR TITLE
Do not rely on replacing `GENERATED_RES` to hook up the plugin

### PR DESCRIPTION
**Pull request by mateuszkwiecinski**
Prefer using `addGeneratedSourceDirectory` API instead

See https://github.com/usefulness/easylauncher-gradle-plugin/issues/382 and  https://issuetracker.google.com/issues/237421684 for more details

### CODE REVIEW

File: easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt [hunk 1]
```diff
@@ -125,3 +142,5 @@ class EasyLauncherPlugin : Plugin<Project> {
             ribbonBuildTypes.filter { it.name == variant.buildType }
     }
 }
+
+private object DUPA : Artifact.Single<Directory>(ArtifactKind.DIRECTORY, Category.GENERATED), Artifact.Replaceable
```
Comment (brudaswen) [inline easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt:146]: Unused?

**Review by brudaswen — COMMENTED** [review]
Looks good. 🦫 Will test tomorrow. But some first comments:

**Comment by brudaswen** [comment]
Just tested with 7.4. Looks good. Great work.

File: easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt [hunk 2]
```diff
@@ -104,11 +107,25 @@ class EasyLauncherPlugin : Plugin<Project> {
                         it.minSdkVersion.set(variant.minSdkVersion.apiLevel)
                     }
 
-                    variant
-                        .artifacts
-                        .use(task)
-                        .wiredWith(EasyLauncherTask::outputDir)
-                        .toCreate(InternalArtifactType.GENERATED_RES)
+                    val apgVersion = androidComponents.pluginVersion
+                    if (apgVersion >= AndroidPluginVersion(7, 4).beta(1)) { // https://issuetracker.google.com/issues/237303854
```
[the remark this answers is not in the document]
Comment (mateuszkwiecinski) [reply easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt]: Ah, right! The `res` property was added in `beta01` ([source](https://issuetracker.google.com/issues/237303854#comment4)), but it seems to be working properly starting from beta02 ([source](https://issuetracker.google.com/issues/237421684#comment16)) 🤷

File: easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt [hunk 3]
```diff
@@ -104,11 +107,25 @@ class EasyLauncherPlugin : Plugin<Project> {
...
+                    } else if (apgVersion < AndroidPluginVersion(7, 3).alpha(1)) {
+                        val generatedResDir = buildDir.resolve("generated/easylauncher/res/${variant.name}")
+                        task.configure { it.outputDir.set(generatedResDir) }
+                        project.afterEvaluate {
+                            val android = extensions.getByName("android") as BaseExtension
+                            android.sourceSets.getByName(variant.name).res.srcDir(generatedResDir)
+                            tasks.named("generate${capitalisedVariantName}Resources") { it.dependsOn(task) }
+                        }
+                    } else {
+                        // has side-effects, but "works". @see: https://github.com/usefulness/easylauncher-gradle-plugin/issues/382
```
[the remark this answers is not in the document]
Comment (mateuszkwiecinski) [reply easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt]: I'll consider adding the log later. This block of code is called on each variant, and we'd spam the console with duplicated messages. Most probably it'll have to happen once, when plugin gets applied? or a task is invoked? It also would be nice to have an ability to disable the log message. I'll see if I manage to add this in a day or two and if not then I'll make a release with what's on the branch currently, relying on the fact people will have to read release notes 🤷

[the remark this answers is not in the document]
Comment (brudaswen) [reply easylauncher/src/main/kotlin/com/project/starter/easylauncher/plugin/EasyLauncherPlugin.kt]: Sounds good!

**Pull request merged**